### PR TITLE
Make pending local echoes sticky, take 2

### PR DESCRIPTION
**Pull request by bnjbvr**
Based on https://github.com/matrix-org/matrix-rust-sdk/pull/2172.

Adds more test expectations, and fix the test issue. Some place in the code assumed that if there's a local and remote echo for the same event, then the local must be placed *before* the remote echo. Now it's not the case anymore, so we have to manually look for a remote echo before updating a local echo to remote; it's not ideal in terms of performance, so this will have to be revisited soon.

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs [hunk 1]
```diff
@@ -799,72 +799,93 @@ impl<'a> TimelineEventHandler<'a> {
...
+
+                    // In more complex cases, remove the item and day
+                    // divider (if necessary) before re-adding the item.
+                    trace!("Removing local echo or duplicate timeline item");
+                    self.items.remove(idx);
+
+                    assert_ne!(
+                        idx, 0,
+                        "there is never an event item at index 0 because \
                              the first event item is preceded by a day divider"
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:812]: Inconsistent indentation

File: crates/matrix-sdk-ui/src/timeline/event_handler.rs [hunk 2]
```diff
@@ -799,72 +799,93 @@ impl<'a> TimelineEventHandler<'a> {
...
+                    // Pre-requisites for removing the day divider:
+                    // 1. there is one preceding the old item at all
+                    if self.items[idx - 1].is_day_divider()
                             // 2. the item after the old one that was removed
                             //    is virtual (it should be impossible for this
                             //    to be a read marker)
                             && self
                                 .items
                                 .get(idx)
                                 .map_or(true, |item| item.is_virtual())
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/event_handler.rs:824]: I guess the comment is confusing rustfmt here, the indentation of these linse should be reduced.

File: crates/matrix-sdk-ui/src/timeline/inner.rs [hunk 3]
```diff
@@ -496,25 +496,20 @@ impl<P: RoomDataProvider> TimelineInner<P> {
...
-        let Some((idx, item)) = result else {
-            // Event isn't found at all.
-            warn!("Timeline item not found, can't add event ID");
-            return;
-        };
-
-        let Some(local_item) = item.as_local() else {
+        // The local echoes are always at the end of the timeline, we must first make
+        // sure the remote echo hasn't showed up yet.
+        // TODO this is expensive to use `rfind_event_item` in that case!
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/inner.rs:501]: Do you mean it's more expensive here than elsewhere? I think it should be fine, we just need to eventually add some ways to control the maximum size of the timeline (there's really no reason to hold many hundrets or thousands of events in memory).

File: crates/matrix-sdk-ui/src/timeline/tests/edit.rs [hunk 4]
```diff
@@ -40,8 +40,9 @@ async fn live_redacted() {
     timeline
         .handle_live_redacted_message_event(*ALICE, RedactedRoomMessageEventContent::new())
         .await;
-    let _day_divider = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
-    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    let _day_divider =
+        assert_next_matches!(stream, VectorDiff::Insert { index:0,  value } => value);
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/edit.rs:44]:
```suggestion
        assert_next_matches!(stream, VectorDiff::Insert { index: 0, value } => value);
```

File: crates/matrix-sdk-ui/src/timeline/tests/edit.rs [hunk 5]
```diff
@@ -71,9 +72,9 @@ async fn live_sanitized() {
         )
         .await;
 
-    let _day_divider = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    let _day_divider = assert_next_matches!(stream, VectorDiff::Insert { index:0, value } => value);
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/edit.rs:75]:
```suggestion
    let _day_divider = assert_next_matches!(stream, VectorDiff::Insert { index: 0, value } => value);
```

File: crates/matrix-sdk-ui/src/timeline/tests/edit.rs [hunk 6]
```diff
@@ -152,8 +153,9 @@ async fn aggregated_sanitized() {
     });
     timeline.handle_live_event(Raw::new(&ev).unwrap().cast()).await;
 
-    let _day_divider = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
-    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    let _day_divider =
+        assert_next_matches!(stream, VectorDiff::Insert { index: 0,  value } => value);
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/edit.rs:157]:
```suggestion
        assert_next_matches!(stream, VectorDiff::Insert { index: 0, value } => value);
```

File: crates/matrix-sdk-ui/tests/integration/timeline/echo.rs [hunk 7]
```diff
@@ -217,7 +217,7 @@ async fn dedup_by_event_id_late() {
                 .set_body_json(&json!({ "event_id": event_id }))
                 // Not great to use a timer for this, but it's what wiremock gives us right now.
                 // Ideally we'd wait on a channel to produce a value or sth. like that.
-                .set_delay(Duration::from_millis(100)),
+                .set_delay(Duration::from_millis(1000)),
```
Comment (jplatte) [inline crates/matrix-sdk-ui/tests/integration/timeline/echo.rs:220]: Oh, this is my change, right? We should revert it.

**Review by jplatte — APPROVED** [review]
Thanks for taking care of this! We really need to move day divider logic out of the timeline event handling logic...

**Comment by jplatte** [comment]
One edge case I'm now wondering about: when starting with an empty timeline and doing a back-pagination, but then sync events arrive before the paginated events, what happens? I think we might put those events *before* the loading indicator, which is obviously wrong.
I think this won't happen in EX where we always wait for a token before paginating, but it should still be handled correctly.

File: crates/matrix-sdk-ui/src/timeline/inner.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/inner.rs:501]: Now we're doing two `rfind_event_item` for each `update_event_send_state` while we were doing one before, and the one for the remote_echo is almost guaranteed to not succeed most of the time, thus iterating over all the elements. We can figure something out later :-)

**Comment by bnjbvr** [comment]
I've slightly changed strategy so we don't have to update all the tests: if we're inserting to the back of the array, then I'm using `push_back` instead of `insert`, which is nicer for the API users too (otherwise they'd have to reconcile that it's a push_back themselves, if they want to distinguish between insert and push_back). It's nicer overall, requires fewer changes, and helps us see in the tests which are *really* changed by the code in the PR.

> One edge case I'm now wondering about: when starting with an empty timeline and doing a back-pagination, but then sync events arrive before the paginated events, what happens? I think we might put those events before the loading indicator, which is obviously wrong.
I think this won't happen in EX where we always wait for a token before paginating, but it should still be handled correctly.

Love the smell of a good future follow-up in the morning...
